tests/test_requestor.py

```diff
@@ -97,22 +98,150 @@ def test_malformed_response(self, reqmock):
+        assert f'client_id={analyzere.oauth_client_id}' in request_text
+        assert f'client_secret={analyzere.oauth_client_secret}' in request_text
+        assert f'scope={analyzere.oauth_scope}' in request_text
+
+    def _make_first_request_with_token(self, reqmock, mock_token, expires_in):
+        reqmock.get(self.api_url, status_code=200)
+
+        reqmock.post(analyzere.oauth_token_url, text=f'{{"access_token": "{mock_token}", "expires_in": {expires_in}}}')
+
+        request_raw('get', self.api_path)
+
+        # One POST to token URL, one GET to API with bearer auth header
+        assert reqmock.call_count == 2
+
+        assert reqmock.request_history[0].url == analyzere.oauth_token_url
+        assert reqmock.request_history[0].method == 'POST'
+        self._validate_token_params(reqmock.request_history[0].text)
+
+        assert reqmock.request_history[1].url == self.api_url
+        assert reqmock.request_history[1].method == 'GET'
+        assert reqmock.request_history[1].headers['Authorization'] == f'Bearer {mock_token}'
+
+    def test_client_credentials_authentication(self, reqmock):
+        mock_token = 's000'
+        self._make_first_request_with_token(reqmock, mock_token, 3600)
+        call_count_after_first_request = reqmock.call_count
+
+        # Make another request, and given the hour-long token expiry, expect no further calls to the token endpoint
+        request_raw('get', self.api_path)
+        assert reqmock.call_count == call_count_after_first_request + 1
+        assert reqmock.last_request.url == 'https://api/bar'
+
+        # Expect the original token to be re-used
+        assert reqmock.last_request.headers['Authorization'] == f'Bearer {mock_token}'
+
+    def test_client_credentials_authentication_refresh(self, reqmock):
+        first_mock_token = 's001'
+        self._make_first_request_with_token(reqmock, first_mock_token, 1)
+        call_count_after_first_request = reqmock.call_count
+
+        # Wait for token to expire
+        time.sleep(1)
+
+        # Mock a second token and make another request
+        second_mock_token = 's002'
+        reqmock.post(analyzere.oauth_token_url, text=f'{{"access_token": "{second_mock_token}", "expires_in": 1}}')
+        request_raw('get', self.api_path)
+
+        # Expect additional POST to token URL given the token expiry
+        assert reqmock.call_count == call_count_after_first_request + 2
+
+        assert analyzere.oauth_token_url in reqmock.request_history[2].url
+        assert reqmock.request_history[2].method == 'POST'
+        self._validate_token_params(reqmock.request_history[2].text)
+
+        assert reqmock.request_history[3].url == self.api_url
+        assert reqmock.request_history[3].method == 'GET'
+
+        # Expect to see the new token
+        assert reqmock.request_history[3].headers['Authorization'] == f'Bearer {second_mock_token}'
+
+    def test_retry_on_401(self, reqmock):
```

Comment:
Just curious, when will we get a 401 status code? Like what triggers it?

Reply:
> needing to handle the case where an API call fails if an access token expires before the expected time. Access tokens can expire for many reasons, such as the user revoking an app, or if the authorization server expires all tokens when a user changes their password.

https://www.oauth.com/oauth2-servers/making-authenticated-requests/refreshing-an-access-token/